trace_filter.py

```diff
@@ -0,0 +1,114 @@
+import numpy as np
+from math import radians, cos, sin, asin, sqrt
+
+# Configurable constants for filtering
+MINIMUM_MEAN_SPEED = 10  # km / h
+MINIMUM_TOTAL_TIME = 120  # seconds, can travel a couple edges
+MINIMUM_TOTAL_DISTANCE = 1000  # meters, about a few city blocks
+MAXIMUM_TIME_BETWEEN_ADJACENT_POINTS = 5  # seconds, we need high granularity to make accurate map matches
+MAXIMUM_SPEED_BETWEEN_ADJACENT_POINTS = 160  # km / h, to weed out poor measurements and trains
+MAXIMUM_POOR_MEASUREMENTS_PERCENT = 0.25  # 25%, max %age of traces that are marked as poor measurements
+
+
+def run(trace_data: list[list[dict]]) -> list[list[dict]]:
+    """
+    Performs simple filters on trace_data. A list of trace data will only be accepted if:
+    - Total time of sequence exceeds MINIMUM_TOTAL_TIME
+    - There are no out-of-order timestamps, i.e. all points are sequential
+    - If time between adjacent points is too long, i.e. adjacent time deltas more than
+        MAXIMUM_TIME_BETWEEN_ADJACENT_POINTS, increment 'num_poor_measurements' counter
+    - If speed between adjacent points is too fast, i.e. adjacent speeds more than
+        MAXIMUM_SPEED_BETWEEN_ADJACENT_POINTS, increment 'num_poor_measurements' counter
+    - The 'num_poor_measurements' counter is too high as per MAXIMUM_POOR_MEASUREMENTS_PERCENT
+    - Total distance of sequence exceeds MINIMUM_TOTAL_DISTANCE
+    - Mean speed is above the walking / driving threshold, MINIMUM_MEAN_SPEED
```

Comment:
Here are the filters that this function performs, in order.

Reply:
This function / file is meant to be generic, taking in lists of trace data in the format that Valhalla map matching expects, and returning a list of all trace data that weren't filtered out.